File: coding_skill_py/12_object_pool.py
"""Python Object Pool"""
from typing import List, Generic, TypeVar, Callable

T = TypeVar('T')

class ObjectPool(Generic[T]):
    def __init__(self, factory: Callable[[], T], size: int):
        self.available: List[T] = [factory() for _ in range(size)]
        self.in_use: List[T] = []
    
    def acquire(self) -> T:
        if not self.available:
            return None
        obj = self.available.pop()
        self.in_use.append(obj)
        return obj
    
    def release(self, obj: T):
        if obj in self.in_use:
            self.in_use.remove(obj)
            self.available.append(obj)

class Packet:
    def __init__(self):
        self.id = 0
        self.data = ""

def main():
    print("=== Python Object Pool ===")
    pool = ObjectPool(Packet, 3)
    
    p1 = pool.acquire()
    p1.id = 1
    print(f"Acquired packet {p1.id}")
    
    pool.release(p1)
    print("Released")

if __name__ == "__main__":
    main()

File: coding_skill_cpp/12_object_pool.cpp
/* C++ Object Pool */
#include <iostream>
#include <vector>
#include <memory>
using namespace std;

template<typename T>
class ObjectPool {
    vector<unique_ptr<T>> pool;
    vector<T*> available;
    
public:
    ObjectPool(size_t size) {
        for (size_t i = 0; i < size; ++i) {
            pool.push_back(make_unique<T>());
            available.push_back(pool.back().get());
        }
    }
    
    T* acquire() {
        if (available.empty()) return nullptr;
        T* obj = available.back();
        available.pop_back();
        return obj;
    }
    
    void release(T* obj) {
        available.push_back(obj);
    }
};

struct Packet { int id; string data; };

int main() {
    cout << "=== C++ Object Pool ===" << endl;
    ObjectPool<Packet> pool(3);
    
    Packet* p1 = pool.acquire();
    p1->id = 1;
    cout << "Acquired packet " << p1->id << endl;
    
    pool.release(p1);
    cout << "Released" << endl;
    
    return 0;
}
